Add support for launch templates to Batch Compute Environments

jeancochrane opened the pull request:
## Overview

Allow users to optionally thread in the ID of launch templates that they'd like to use for either CPU- or GPU-enabled jobs.

Closes #11.

## Notes

This approach requires a user to create a launch template separately from this CloudFormation stack and then thread it in. The downside of this approach is that the launch templates have to be managed separately from the CloudFormation stack, and more AWS labor/expertise is required of the user. The upside is that launch templates will be far more flexible than they would be if we tried to create default launch templates in the stack and then allow the user to tweak parameters in them. Since the addition of support for launch templates is intended to add flexibility to the module, I reasoned that the tradeoffs were acceptable for this approach. 

## Testing instructions

### 1. Test creating a stack without launch templates

* Create a stack following the instructions in https://github.com/azavea/raster-vision-aws#deploying-batch-resources
  * Use the `Raster Vision VPC`
  * Use the subnet `subnet-033d2cbe8268c3067`
  * Use the GPU AMI `ami-0d4fd80a4230ab510`
  * Use the CPU AMI `ami-011a85ba0ae2013bf`
  * Leave the values for GPU and CPU launch templates empty
* Confirm that resources build properly
* Tear down your stack

### 2. Test creating a stack with launch templates

* Create a launch template in `EC2 > Launch templates > Create launch template`
  * Under `Advanced details > User data`, add the following cloud-init config:

```
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="==MYBOUNDARY=="

--==MYBOUNDARY==
Content-Type: text/cloud-config; charset="us-ascii"

runcmd:
- echo "hello, world!" > /home/ec2-user/helloworld.txt

--==MYBOUNDARY==--
```

* Create a new stack following the instructions in part 1
  * For the GPU and CPU launch templates, thread in the ID of the launch template you just created
* In your Batch job queue (either CPU or GPU), submit a job with the override command `cat /opt/data/helloworld.txt` and confirm that the job echos the hello world file that your launch template should create
* Delete the stack you created

jeancochrane commented on settings.mk.template:
```diff
@@ -12,3 +12,6 @@ AWS_REGION := us-east-1
 ECR_IMAGE_TAG := latest
+
+# Settings for Terraform
+SUBNET_IDS := [\"subnet-XXXXXXXXXXXXXXXXX\"]
```
The escaped double quotes are necessary here in order to be able to pass this Make variable into Terraform with the `-var` command line flag. I don't love it, but I expect that this config file will be removed entirely in #8 in favor of a standard Terraform variable file.

jeancochrane commented on terraform/batch.tf:
```diff
@@ -1,10 +1,6 @@
 #
-data "aws_iam_instance_profile" "batch_instance_role" {
-  name = "${aws_iam_role.container_instance_ec2.name}"
-}
```
This change effectively reverts https://github.com/azavea/raster-vision-aws/commit/1caf7fd66ef2a675769e398cc246d7790068e129. I couldn't figure out what that commit was trying to fix, but the use of an extra reference here was causing my stack to fail to deploy because the `aws_iam_role.container_instance_ec2` resource didn't reliably exist before this data source was evaluated. Rather than try to wire up an explicit dependency into this data source with `depends_on`, I tried referencing the IAM instance profile directly and things deployed properly for me. @lossyrob may have more context on why  https://github.com/azavea/raster-vision-aws/commit/1caf7fd66ef2a675769e398cc246d7790068e129 was necessary, and if it's still relevant then I think we can get around the deployment race condition issue with a `depends_on` dependency.

jeancochrane replied:
Rob reports the `iam_instance_profile` data source was actually a fix discovered by @hectcastro working through Raster Foundry bugs.

jeancochrane replied:
Here's the related Raster Foundry commit, for future reference: https://github.com/azavea/raster-foundry-deployment/commit/4c7fd3c751f1ac91ea3a6f6deea6df82a8333ded AWS Batch needs a reference to an instance profile, not the role that's contained within that profile. The data source is a quick way to swap in the right reference.

hectcastro commented on terraform/security_group.tf:
```diff
@@ -2,7 +2,7 @@
+  vpc_id = "${data.aws_subnet.first.vpc_id}"
 
   ingress {
     from_port   = 22
```
I recognize that this isn't in scope for the PR, but I think that it would be valuable to include a variable for the `cidr_blocks` of the security group rules defined in this resource. Currently, they are opening up 22 and 6006 to the world, and it would be better to just open these ports to a known range.

jeancochrane replied:
Great point! Opened up #15 to track this.

hectcastro commented on terraform/security_group.tf:
```diff
@@ -31,3 +31,7 @@ resource "aws_security_group" "container_instance" {
   }
 }
+
+data "aws_subnet" "first" {
```
I would consider raising this data resource to the top of the file for better visibility (I feel like I think of them like variable assignments to a degree).

hectcastro commented on terraform/batch.tf:
```diff
@@ -3,6 +3,10 @@
 data "aws_iam_instance_profile" "batch_instance_role" {
   name = "${aws_iam_role.container_instance_ec2.name}"
+
+  depends_on = [
```
Do you run into an error without this? Using the `${}` syntax should automatically create the correct order of operations for these two resources.

jeancochrane replied:
I do, somewhat to my surprise. I was expecting the `name` attribute to create an implicit dependency, too. Here's the error I get when I remove the explicit dependency:

```
data.aws_iam_instance_profile.batch_instance_role: Error getting instance profiles: NoSuchEntity:
Instance Profile rasterVisionTestingContainerInstanceProfile cannot be found.
```

Interestingly, the logs show that the `aws_iam_role.container_instance_ec2` resource gets created before this error gets thrown, which I would expect to prevent the error in the first place.

hectcastro replied:
Wasn't able to find much that identifies this specific issue, but I think keeping the `depends_on` here is fine.

Part of me feels like this wouldn't be happening if we were trying to resolve `aws_iam_role.container_instance_ec2.arn` vs. `aws_iam_role.container_instance_ec2.name`, but we need name here, and any hacks to pursue that theory seem like they'd look a lot worse than `depends_on`.

Commit:
Add support for optional launch templates

Allow users to optionally thread in the ID of launch templates that
they'd like to use for either CPU- or GPU-enabled jobs.

Commit:
Add versioning to launch templates

Allow the user to specify a different version of a launch template for
use in CPU or GPU environments.

jeancochrane commented on cloudformation/template.yml:
```diff
@@ -208,11 +220,41 @@ Parameters:
+    Default: 1
+    Description:
+      (Optional) The version of the EC2 Launch Template to associate with the GPU-based
+      Batch Compute Environment. 
```
I'm not sure how useful it will be to expose versioning for launch templates. On the one hand, exposing the version should hypothetically allow the user to update a stack to read from a new version if they update their template, which seems useful. On the other hand, while testing this I ran into some unexpected behavior where [CloudFormation won't update a custom-named resource](https://aws.amazon.com/premiumsupport/knowledge-center/cloudformation-custom-name/), which happens to include the compute environment in this module. That means that in order to update your stack without recreating it you'd have to do something hacky like rename the compute environment out of band before updating the stack (at which point you'll probably just want to skip it all and update the launch template version for the compute environment out of band).

In the ideal world we could just set this to pull the "default" version of the launch template, but [this blog post points out](https://www.kindlyops.com/knowledge-base/cloudformation-launch-templates/) that if no version number is specified CloudFormation will default to version `1`.

hectcastro replied:
Interesting. For the ECS cluster module in Terraform, there is a magic [`$$Latest`](https://github.com/azavea/terraform-aws-ecs-cluster/blob/9d7ff9af1c2807d55bb369296fc187a3a67d108a/main.tf#L192) that pulls the latest version of the launch template. It is unclear whether this is a magic variable evaluated by AWS, or Terraform itself.

It also looks like `AWS::EC2::LaunchTemplate` has a `LatestVersionNumber` attribute.

Would it be possible to default to the latest if `GpuLaunchTemplateVersion` is not specified?

jeancochrane replied:
Interesting! I couldn't find any official documentation about how to retrieve the latest version, and since it defaults to version 1 I assumed that there wasn't one. I think it's worth giving `$$Latest` a try, however.

jeancochrane commented on cloudformation/template.yml:
```diff
@@ -336,6 +378,13 @@ Resources:
+            -
+              LaunchTemplateId: !Ref GpuLaunchTemplate
+              Version: !Ref GpuLaunchTemplateVersion
+            - !Ref "AWS::NoValue"
```
I'd never seen this syntax before, but `!If` is basically just a ternary operator: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-conditions.html#intrinsic-function-reference-conditions-if

hectcastro replied:
Neither have I. This looks like exactly what we're looking for in Terraform land with regard to ASG `desired_capacity`.

hectcastro reviewed (commented):
Some minor comments, but I'm interested in what motivated you to pursue an approach that passes in the launch template rather than creating it within CloudFormation?

jeancochrane commented:
> I'm interested in what motivated you to pursue an approach that passes in the launch template rather than creating it within CloudFormation?

My main motivation was that launch templates expose a ton of different features (or at least that's the impression I get looking at the docs: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-launchtemplatedata.html) meaning that if we were to create a launch template based on user parameters we would have to drastically narrow the scope of what the template could do. Maybe for RV that's acceptable -- if we only want to support e.g. adjusting the size of the mounted filesystem or adding custom user data, then it would probably make sense to build the launch template as part of this stack -- but since launch templates seemed to me to be a feature for advanced customization, I defaulted to leaving as much control as possible over the launch template to the end user.

hectcastro commented:
Superseded by https://github.com/azavea/raster-vision-aws/pull/16.
